ledger/apply/application_test.go

```diff
@@ -962,6 +963,58 @@ func TestAppCallApplyUpdate(t *testing.T) {
 	a.Equal([]byte{2}, br.AppParams[appIdx].ApprovalProgram)
 	a.Equal([]byte{2}, br.AppParams[appIdx].ClearStateProgram)
 	a.Equal(basics.EvalDelta{}, ad.EvalDelta)
+
+	// check app program len
+	appr := make([]byte, 6050, 6050)
+
+	for i := range appr {
+		appr[i] = 2
+	}
+	appr[0] = 4
+	ac = transactions.ApplicationCallTxnFields{
+		ApplicationID:     appIdx,
+		OnCompletion:      transactions.UpdateApplicationOC,
+		ApprovalProgram:   appr,
+		ClearStateProgram: []byte{2},
+	}
+	params = basics.AppParams{
+		ApprovalProgram: []byte{1},
+		StateSchemas: basics.StateSchemas{
+			GlobalStateSchema: basics.StateSchema{NumUint: 1},
+		},
+		ExtraProgramPages: 1,
+	}
+	h = transactions.Header{
+		Sender: sender,
+	}
+
+	b.balances = make(map[basics.Address]basics.AccountData)
+	cbr = basics.AccountData{
+		AppParams: map[basics.AppIndex]basics.AppParams{appIdx: params},
+	}
+	cp = basics.AccountData{
+		AppParams: map[basics.AppIndex]basics.AppParams{appIdx: params},
+	}
+	b.balances[creator] = cp
+	b.appCreators = map[basics.AppIndex]basics.Address{appIdx: creator}
+
+	// check program len check doesn't happen in current consensus proto version
+	b.SetProto(protocol.ConsensusV28)
+	proto = b.ConsensusParams()
+	ep.Proto = &proto
+
+	b.pass = true
+	err = ApplicationCall(ac, h, &b, ad, &ep, txnCounter)
+	a.NoError(err)
+
+	//check program len check happens in future consensus proto version
+	b.SetProto(protocol.ConsensusFuture)
+	proto = b.ConsensusParams()
+	ep.Proto = &proto
+
+	b.pass = true
+	err = ApplicationCall(ac, h, &b, ad, &ep, txnCounter)
+	a.Contains(err.Error(), "updateApplication app programs too long")
```

Comment:
We want to confirm that it IS possible to have a 3k approval program here.  Currently, you can not update to 3k, even though your extra pages = 1 when you created it (which is what I think this is testing).

Reply:
this test is for validating this added check lines are actually invoked, https://github.com/algorand/go-algorand/blob/af3858cf92829aa8f6eb2ffcedd4a03c0b820c88/ledger/apply/application.go#L203-L209. 

But, yes, I need to add another test to validate that it is possible to have a 3K approval program.